chore: enforce tenant limits on the frontend

rdubrock opened the pull request:
I'm a little unsure where the best place to gate access based on limits is. I definitely don't want it to be how it currently is (when you submit a form after filling it out), but there are other options than what I've done. Opinions welcome.

Fixes https://github.com/grafana/synthetic-monitoring/issues/49

No limits reached:
![Screenshot 2024-04-03 at 14 07 38](https://github.com/grafana/synthetic-monitoring-app/assets/8377044/5f3600f2-7968-4475-9fa2-97aa1654ffb2)

Overall limit reached:
![Screenshot 2024-04-03 at 13 13 38](https://github.com/grafana/synthetic-monitoring-app/assets/8377044/8a4d5ae0-2dc3-4333-97a6-aa0bc9d1db0c)

Scripted limit reached:
![Screenshot 2024-04-03 at 13 12 39](https://github.com/grafana/synthetic-monitoring-app/assets/8377044/ac2cd80d-7d61-4508-9fbf-c2d55ab5d1e9)

rdubrock commented on src/components/LabelField/LabelField.tsx:
```diff
@@ -17,23 +18,28 @@ type FormWithLabels = {
-  let description = `Custom labels to be included with collected metrics and logs.`;
-
+  let description = '';
   if (limit) {
```
I don't love this logic. Basically we're saying "some limits come from the api, and by not providing a limit prop we assume those are the ones you're looking for". We need that case for handling private probe label limits. Maybe just changing the name of the prop would help.

rdubrock replied:
I've changed it to something that makes me happier :)

ckbedwell replied:
Nit: I'm not that precious about this but I'd extract that function to clean up the render method and reduce some responsibilities for each function.

```
 // in the render method
  const limit = getLimit(labelDestination, limits);
  const description = getDescription(labelDestination, limit, limits.maxAllowedLogLabels); 
```

```
// defined separately
function getLimit(labelDestination: LabelFieldProps['labelDestination'], limits: ListTenantLimitsResponse) {
  if (labelDestination === 'check') {
    return 3;
  }

  if (limits.maxAllowedMetricLabels) {
    return limits.maxAllowedMetricLabels;
  }

  return 10
}

function getDescription(labelDestination: LabelFieldProps['labelDestination'], limit: number, logLabelLimit: number) {
  if (labelDestination === 'check') {
    return `Custom labels to be included with collected metrics and logs. You can add up to ${limit}.`
  }

  return `Custom labels to be included with collected metrics and logs. You can add up to ${limit}. If you add more than ${logLabelLimit} labels, they will potentially not be used to index logs, and rather added as part of the log message.`
}
```

ckbedwell reviewed (changes requested):
Left some comments. More general ones:

I'd account for the loading / error states in the UI for the `LabelField`. I don't love that if the api call fails on the client for whatever reason that we fallback to the defaults that might differ to each stack for whatever reason.

I know it is minor but I'd rather remove the LabelField element and have a retry button about trying to refetch the tenant limits so we can decipher the amount of labels they can have rather than us guessing. This would be a great use case for adding a Faro custom event to see if we ever show / anybody uses the refetch button in this scenario.

I left a comment about maybe disabling the Scripted / Multihttp card rather than removing it from the UI and leaving a message. After a while users will probably get sick of looking at it if they know they have reached the limit but want to create other checks.

Maybe similar when the check creation is reached but I think the more general alert works in this case as I can't see a use-case of continually coming back to this page if you know you have reached the overall limit.

Also, on this point, we should add messaging and disable the form submit on the check pages themselves. Some clever sausage might think to go to the route directly to see if they can submit a check (I'm assuming the api will reject it?) so we can save them the time and show them the message / disable the UI upfront for them.

rdubrock commented:
> I know it is minor but I'd rather remove the LabelField element and have a retry button about trying to refetch the tenant limits so we can decipher the amount of labels they can have rather than us guessing.

Yeah I waffled on this. I'd rather not block them doing what they want. I think maybe better to show a warning with the retry button but still default. The defaults are the _minimum_ quantity of custom labels available, and it's only going to be a tiny minority of checks that need more than the minimum anyways. So I'm more worried about cutting off the path to getting the check submitted than the path to attach more than the minimum quantity of labels

rdubrock commented on src/components/ChooseCheckType.test.tsx:
```diff
@@ -0,0 +1,91 @@
+
+it('shows error alert when scripted check limit is reached', async () => {
+  // @ts-ignore
+  config.featureToggles[FeatureName.ScriptedChecks] = true;
```
We could do, I don't have feelings on hiding the card vs restyling the card and making it disabled. I originally thought I'd just disable the card, but we don't have great design patterns for disabled cards actually looking disabled (in grafana ui they just turn off the pointer events and make no visual distinction, which is just confusing imo).

I think that the test passing with the flag off is still kinda valid? Like I'm now sure _how_ a user who doesn't have access to scripted checks in the UI would hit the limits, but I don't see why we'd want the experience to be different in that scenario. Like if they were using terraform or something and the flag states were out of sync with each other, I don't think it's bad to notify the user they're at their limit.

ckbedwell replied:
>We could do, I don't have feelings on hiding the card vs restyling the card and making it disabled. I originally thought I'd just disable the card, but we don't have great design patterns for disabled cards actually looking disabled (in grafana ui they just turn off the pointer events and make no visual distinction, which is just confusing imo).

Yeah, fair point. 👍 

>I think that the test passing with the flag off is still kinda valid?

It's just a bit of a red herring. The thing is the tests double as documentation so if I were to read this in a few months' time and notice the feature toggle had been explicitly set in the test but we don't account for that in the code I'd think our implementation is wrong and might fix it. I'd prefer this test if we either explicitly set the feature toggle to false with a comment or we just removed it altogether as it has no affect.

I don't know if this would ever be a thing but I was wondering if we would ever set their limit to 0 scripted checks then they would see this regardless of if they had access or knew scripted checks were a thing.

Also, we want to remove `config.featureToggles[FeatureName.ScriptedChecks] = true;` this line in this test too as it is still going to pollute other tests :)

rdubrock commented on src/components/CheckForm/FormLayouts/CheckTCPLayout.tsx:
```diff
@@ -39,7 +39,7 @@ export const CheckTCPLayout = () => {
       </FormLayout.Section>
       <FormLayout.Section label="Advanced options" fields={[`labels`, `settings.tcp.ipVersion`]}>
-        <LabelField<CheckFormValuesPing> />
+        <LabelField<CheckFormValuesPing> labelDestination="probe" />
```
Whoops, Friday brain on that one!

rdubrock commented:
Error state: 

![Screenshot 2024-04-08 at 11 46 33](https://github.com/grafana/synthetic-monitoring-app/assets/8377044/be83b112-3e78-4078-9ba1-a40c5cf4744c)

rdubrock commented on src/components/CheckEditor/CheckEditorNew.test.tsx:
```diff
@@ -257,4 +257,50 @@ describe('new checks', () => {
+
+      const errorAlert = await screen.findByText(/Maximum number of scripted checks reached/, { exact: false });
+      expect(errorAlert).toBeInTheDocument();
+      const submitButton = await screen.findByRole('button', { name: 'Save' });
```
I know we just got rid of the `byRole` selectors, but `byText` returns the span inside the button, so I can't assert whether it's disabled or not

ckbedwell replied:
I'm happy to let them sneak back in and if we start seeing timeouts reappear we can re-address this. We have some helper functions so there is an avenue to abstract some of this (I think how I handled this in other tests is I got the parentNode from the getByText selector... 🤮)

ckbedwell commented on src/datasource/DataSource.ts:
```diff
@@ -352,6 +353,14 @@ export class SMDataSource extends DataSourceApi<SMQuery, SMOptions> {
       });
   }
 
+  async getTenantLimits(): Promise<ListTenantLimitsResponse> {
+    return firstValueFrom(
+      getBackendSrv().fetch({ method: 'GET', url: `${this.instanceSettings.url}/sm/tenant/limits` })
+    ).then((res: any) => {
+      return res.data;
+    });
+  }
```
Just going to bump this comment in case you missed it. For all intents and purposes of how we use this function it is virtually the same as what you had except gets rid of what could be problematic explicit any types if `@grafana/runtime` was updated and changed their behaviour for any reason.

rdubrock replied:
Whoops missed this one

ckbedwell reviewed (commented):
Two minor things then this is good to go!

rdubrock commented:
>  I'd prefer this test if we either explicitly set the feature toggle to false with a comment or we just removed it altogether 

Don't want to dig in my heels too much, but setting the flag to false to me is testing that the feature flag is hiding the card (which we also have a test for). Setting the flag to true is actually the scenario users are going to experience, so I'd rather test what's going to happen in the wild. I added an explanatory note. Not a hill I'm trying to die on, though, if that doesn't cut it we can delete it no problem

ckbedwell reviewed (approved):
LGTM! 🚀
